`thread::spawn`, Send, Arc, Mutex (#119)
ferrous-systems/rust-training: training-slides/src/deconstructing-send-arc-mutex.md

<!-- markdownlint-disable MD031 MD033 MD037 -->
# Deconstructing Send, Arc, and Mutex

> ## `tread::spawn` Function

Reviewer: ```suggestion
## `thread::spawn` Function
```

```rust ignore
pub fn spawn<F, T>(f: F) -> JoinHandle<T>
where
    F: FnOnce() -> T,
    F: Send + 'static,
    T: Send + 'static,
{
    // ...
}
```

## Quick Primer on Rust Closures

* 3 categories of data

>   * data the closure *closes over* / *captures*: **Upvars**

Reviewer: oooh, TIL. We should call them that in the closures deck.

    * convenient compiler terminology
    * not represented by closure type signature
  * parameters
  * returned value

```rust ignore
let upper_threshold = 20;
let outliers: Vec<_> = data.iter().copied().filter(|n| -> bool {
    // `n` is a parameter, `upper_threshold` is an *upvar*
    n >= upper_threshold
}).collect();
```

## Spawn closure type

* `F: FnOnce() -> T`
  * closure doesn't accept any parameters
  * closure can *consume upvars* ("FnOnce")
* `F: Send + 'static`
  * applies to *upvars*
* `T: Send + 'static`
  * applies to returned value

## `T: 'static`

> One of two cases:

Reviewer: Perhaps:

```suggestion
Two allowable options:
```

* the type doesn't have any references inside ("Owned data")
  * `struct User { name: String }`
* the references inside the type are `'static`
  * `struct Db { connection_string: &'static str }`

## Why `F: 'static` and `T: 'static`?

> * applies to data passed from parent thread to child thread or vise versa

Reviewer: ```suggestion
* applies to data passed from parent thread to child thread or vice-versa
```

* prevents passing references to local variables
  * one thread can finish before the other and such references may become invalid

>   * `+ 'static` lets borrow checker detect these situations at compile time

Reviewer: ```suggestion
  * `+ 'static` avoids this by ensuring any references point to data that has the static lifetime (i.e. that lives forever)
```

## `T: Send`

`pub unsafe auto trait Send { }`

* `auto` means all types get this trait automatically
  * opt-out instead of opt-in
* various types in standard library implement `Send` or `!Send`
* `unsafe` means you have to put `unsafe` keyword in front of `impl` when implementing `Send` or `!Send`

>   * precaution measure

Reviewer: ```suggestion
  * precautionary measure
```

## Why would one implement `Send` or `!Send`

> * Rust pointers (`*const T`, `*mut T`, `NonNull<T>`) are `!Send`

Reviewer: ~~Also, `UnsafeCell` is !Send, if I remember right.~~

It's Send but !Sync.

  * Use-case: what if the pointer comes from FFI library that assumes that all functions using this pointer are called from the same thread?
* `Arc` has a `NonNull<..>` inside and becomes `!Send` automatically
  * to override this behavior `Arc` explicitly implements `Send`

## `Send` in `thread::spawn` Function

`F: Send` and `T: Send` means that all data traveling from the parent thread to child thread has to be marked as `Send`

> * Rust compiler has to inherent knowledge of threads, but the use of marker traits and lifetime annotations let the type / borrow checker prevent data race errors

Reviewer: ```suggestion
* Rust compiler has no inherent knowledge of threads, but the use of marker traits and lifetime annotations let the type / borrow checker prevent data race errors
```

* `log.push(line.len());`

## "Dream" API

```rust ignore
fn handle_client(stream: net::TcpStream, log: &mut Vec<usize>) -> Result<(), io::Error> {
    // ...
    for line in ... {
        log.push(line.len());
        // ...
    }
    Ok(())
}

fn main() -> Result<(), io::Error> {
    let mut log = vec![];

    for stream in listener.incoming() {
        // ...
        thread::spawn(|| {
            let _ = handle_client(stream, &mut log);
        });
    }
    Ok(())
}
```

## Errors

<pre>
error[E0373]: closure may outlive the current function, but it borrows `log`,
which is owned by the current function
  --> src/main.rs:26:23
   |
26 |         thread::spawn(|| {
   |                       ^^ may outlive borrowed value `log`
27 |             let _ = handle_client(stream, &mut log);
   |                                                --- `log` is borrowed here
   |
note: function requires argument type to outlive `'static`
</pre>

## Life-time problem

Problem:

* local data may be cleaned up prematurely

Solution:

* move the decision when to clean the data from compile-time to runtime
  * use reference-counting

## Attempt 1: `Rc`

* `let mut log = Rc::new(vec![]);`

> * `let mut thread_log = log.clone()` now does't clone the data, but simply increases the reference count

Reviewer: ```suggestion
* `let mut thread_log = log.clone()` now doesn't clone the data, but simply increases the reference count
```

  * both variables now have *owned* type, and satisfy `F: 'static` requirement

```text
error[E0277]: `Rc<Vec<usize>>` cannot be sent between threads safely
```

## `Rc` in Rust Standard Library

* uses `usize` for reference counting
* explicitly marked as `!Send`

```rust ignore
pub struct Rc<T> {
    ptr: NonNull<RcBox<T>>,
}

impl<T> !Send for Rc<T> {}

struct RcBox<T> {
    strong: Cell<usize>,
    weak: Cell<usize>,
    value: T,
}
```

## `Arc` in Rust Standard Library

* uses `AtomicUsize` for reference counting
* explicitly marked as `Send`

```rust ignore
pub struct Arc<T> {
    ptr: NonNull<ArcInner<T>>,
}

impl<T> Send for Arc<T> {}

struct ArcInner<T: ?Sized> {
    strong: atomic::AtomicUsize,
    weak: atomic::AtomicUsize,
    data: T,
}
```

## `Rc` vs `Arc`

* `Arc` uses `AtomicUsize` for reference counting
  * slower
  * safe to increment / decrement from multiple threads

> * With the help of marker trait `Send` and trait bounds on `thread::spawn` Rust compiler *forces* you to use the correct type

Reviewer: Perhaps:

```suggestion
* With the help of marker trait `Send` and trait bounds on `thread::spawn`, the compiler *forces* you to use the correct type
```

        guard.push(line.len());
        // line above expands to:
        // Vec::push(<MutexGuard<'_, _> as DerefMut>::deref_mut(&mut guard), line.len());
        writeln!(writer, "{}", line)?;
        writer.flush()?;
    }
}
```

## `Mutex` locking and unlocking

* we `lock` the mutex for exclusive access to underlying data at runtime
* old C APIs used a pair of functions to lock and unlock the mutex
* `MutexGuard` does unlocking automatically when is dropped
  * time between guard creation and drop is called *critical section*

## Lock Poisoning

* `MutexGuard` in its `Drop` implementation checks if it is being dropped normally or during a `panic` unwind
  * in later case sets a poison flag on the mutex
* calling `lock().unwrap()` on a poisoned Mutex causes `panic`
  * if the mutex is *"popular"* poisoning can cause many application threads to panic, too.
* poisoning API is *problematic*
  * `PoisonError` doesn't provide information about the panic that caused the poisoning
  * no way to recover and revive the mutex (stays poisoned forever)
    * `PoisonError::into_inner` *can* produce a guard even for poisoned mutexes

> ## Critical Section "Hygiene"

Reviewer: This slide didn't fit on my screen :(

* keep it short to reduce the window when mutex is locked
* avoid calling functions that can panic
* using a named variable for Mutex guard helps avoiding unexpected temporary lifetime behavior

```rust ignore
fn handle_client(..., log: &Mutex<Vec<usize>>) -> ... {
    for line in ... {
        {
            let mut guard: MutexGuard<'_, Vec<usize>> = log.lock().unwrap();
            guard.push(line.len());
        } // critical section ends here, before all the IO
        writeln!(writer, "{}", line)?;
        writer.flush()?;
    }
}
```
<p>&nbsp<!-- run-button placeholder --></p>

* `drop(guard)` also works, but extra block nicely highlights the critical section

## Lessons Learned

> * careful use of traits and trait boundaries lets Rust compiler detect problematic multithreading code at compile time

Reviewer: ```suggestion
* careful use of traits and trait boundaries lets the compiler detect problematic multi-threading code at compile time
```

Or *the Rust compiler*
